### ISSUE TITLE
Performance difference between interpreter & unit tests

### ISSUE THREAD

**Issue by intoverflow** [valida-xyz/valida#33]
I tried the instructions from the README and was a bit surprised by the results.

When running the interpreter, fib(25) takes 104 seconds:

```console
$ echo -n '\x19' | time ./target/release/valida basic/tests/data/fibonacci.bin
Command terminated by signal 9
104.76user 16.74system 2:01.90elapsed 99%CPU (0avgtext+0avgdata 62092988maxresident)k
38656inputs+0outputs (194major+16075827minor)pagefaults 0swaps
```

The README suggested piping to hexdump, but doing so produces no output, so I elided that command.

I don't know how many cycles of execution this takes, but looking at the assembly, seems like 10-ish instructions per Fib iteration? So this lands somewhere around 250 instructions / 104 seconds = just a few Hz. That doesn't seem right; what's going on?

The unit tests, however, complete *much* quicker:

```console
$ cargo test --release

...

     Running `/home/tcarstens/valida/valida/target/debug/valida tests/data/fibonacci.bin`
test run_fibonacci ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.10s

     Running tests/test_prover.rs (target/release/deps/test_prover-6787c50d144a7a27)

running 1 test
test prove_fibonacci ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

Using the rough time for the interpreter unit test (0.1 seconds), this comes in around 250 instructions / 0.1 seconds = 2.5 kHz, which also doesn't seem right. (The prover completed in fewer than 0.00 seconds, so I can't do a similar comparison. Why would proving be faster, though?)

Why is the interpreter so much slower than the unit test? And what's the right way to calculate the effective clockspeed?

**Comment by intoverflow** [comment]
(I'm also not sure why the interpreter is dying with signal 9, but will investigate later)

**Comment by intoverflow** [comment]
Ok, I investigated the sig9 death. I think the process is running out of RAM:

```console
$ echo -n '\x19' | time -f "%e %M" ./target/release/valida basic/tests/data/fibonacci.bin
Command terminated by signal 9
122.53 62261496
```

This indicates 62,261,496 KB resident memory usage, which is pretty much all I've got on this machine.

**Comment by intoverflow** [comment]
Hi Max! Thank you for the quick reply.

I'm currently on the latest commit:

```console
$ git rev-parse HEAD
cd150291482e4909f01676c3c7efeaf5f7f827ac
```

I ran the unit tests with `cargo test --release`.

I'm running on x86, not Mac; not sure if that's relevant

**Comment by intoverflow** [comment]
I also tried updating the interpreter unit test to compute fib(250000), but found that this broke the unit test:

Test output:

```console
     Running `/home/tcarstens/valida/valida/target/debug/valida tests/data/fibonacci.bin`
test run_fibonacci ... FAILED

failures:

---- run_fibonacci stdout ----
thread 'run_fibonacci' panicked at 'assertion failed: `(left == right)`
  left: `3256916931`,
 right: `243736795`', basic/tests/test_interpreter.rs:44:5
```

The edit (impacts [this line](https://github.com/valida-xyz/valida/blob/main/basic/tests/test_interpreter.rs#L33)):

```console
$ git diff
diff --git a/basic/tests/test_interpreter.rs b/basic/tests/test_interpreter.rs
index a674c5e..c5ed1f0 100644
--- a/basic/tests/test_interpreter.rs
+++ b/basic/tests/test_interpreter.rs
@@ -30,7 +30,7 @@ fn run_fibonacci() {
         .expect("failed to execute process");

     // Write the desired Fibonacci number to stdin
-    let number = 25;
+    let number = 250000;
     let stdin = child.stdin.as_mut().expect("failed to get stdin");
     stdin.write_u32::<LittleEndian>(number).unwrap();
```

**Comment by maxgillett** [comment]
I fixed the issue with fib(250000) (and hopefully any other fib values) in the latest commit (36bb3ab753924eb80497379dc7c9159dae445df5). The output was being cast to ASCII characters, so it wasn't printing byte values greater than 127 correctly.

I tried to reproduce the other issue on my Linux machine (x86), and got the expected behavior. Can you share more details about your machine?

**Comment by intoverflow** [comment]
I'm on Linux/x86, specifically PopOS on a System76 laptop. 12th Gen Intel(R) Core(TM) i7-12700H.

I pulled the latest `valida`:

```console
tcarstens@pop-os:~/valida/valida$ git rev-parse HEAD
36bb3ab753924eb80497379dc7c9159dae445df5
```

And am using the latest Plonky3:

```console
tcarstens@pop-os:~/valida/Plonky3$ git rev-parse HEAD
c35cda3fa8c18af9c4bd807dac0bef540c34e206
```

With the latest `valida`, I can edit the unit tests and get the expected behavior 🥳! The unit test for non-proving fib(250000) completes in 5.23 seconds (`cargo test --release`), roughly 335kHz assuming 7 cycles per Fib iteration (likely slowed-down by running within a unit test vs native bin).

The interpreter binary still doesn't work, though.

**Comment by intoverflow** [comment]
I figured out the issue: the interpreter is working just fine, but `echo` isn't parsing the given string. I created a PR to update the README: https://github.com/valida-xyz/valida/pull/35

Now that we've got that figured out, I decided to try running the interpreter for fib(250000):

```console
$ printf "\x90\xd0\x03" | time -f "%e %M" ./target/release/valida basic/tests/data/fibonacci.bin | hexdump
0.46 282016
0000000 20db 0e87
0000004
```

This completed in 0.46 seconds, giving an effective clockspeed of 7 * 250000 / 0.64 = 2.7 MHz. Not as fast as your Mac, but still very fast!

Thank you again @maxgillett for your help! This is cool to see
